### PR TITLE
Fix lots of georeferencer issues

### PR DESCRIPTION
**Pull request by nyalldawson**
Before trying to fix the mess that has gotten into, we need a good test suite...!

### CODE REVIEW

**Review by 3nids — APPROVED** [review]

**Comment by roya0045** [comment]
Unsure if this can be helpful but there were some exchanges on https://github.com/qgis/QGIS/pull/46415

The renaming of things is welcome.

**Comment by nyalldawson** [comment]
@roya0045 @gioman I'd love testing of this branch and confirmation that you don't see any remaining issues (aside from those described in https://github.com/qgis/QGIS/pull/46941)

The size of these changes makes me nervous, but at the same time the georeferencer is quite broken in master/3.22 so we don't have any choice but to merge and backport and hope that we can test sufficiently prior to release...!

**Comment by gioman** [comment]
@nyalldawson I tested the windows artifact. The georef operation seems to get stuck (progress bar moves, but like 1%  every 5 minutes) and GCP points are a no show in georef map window:

image

I'm using the data and configs from here https://github.com/qgis/QGIS/issues/46901#issuecomment-1016957555

A few days back I tested the artifact for a now closed patch https://github.com/qgis/QGIS/pull/46524#issuecomment-1024211647 and worked ok (even if I understand that was not the right fix).

**Comment by gioman** [comment]
@nyalldawson GCP did show up when I double clicked somwhere in the GCP table, but that made the image to georef disappear

image

**Comment by nyalldawson** [comment]
@gioman I suspect the points file there is invalid, and has map coordinates for the raster instead of pixels. Creation of the gcp file is fixed here, but I'm not sure we can "auto repair" invalid ones...

**Comment by gioman** [comment]
@nyalldawson ok, I'll try with another dataset.

**Comment by nyalldawson** [comment]
@gioman

It's always possible that I've made the wrong choice here, and that we should be storing the source **coordinates** instead of pixels in the gcp file! I'm not sure what the original intention was, but we always interpret these values as pixels and the heading states "pixelX" so that's why I went that way.

But maybe the heading is what's wrong... 😳

Thinking about a future time when we support vector file georeferencing, it no longer would make sense to use pixels here.

So in short... I'm 50/50, and would value any insights anyone has!

**Comment by gioman** [comment]
@nyalldawson tested again from scratch. The operation seems to work ok. Anyway I have a few observations I will leave here. As a start;

I georeferenced a jpg using a project that had CRS 3763 and picking the GCP on the canvas. In georef configs I choose 3763 as target CRS. The output is added to QGIS with "unknown" CRS (but in the right place).

**Comment by gioman** [comment]
> @nyalldawson tested again from scratch. The operation seems to work ok. Anyway I have a few observations I will leave here. As a start;
> 
> I georeferenced a jpg using a project that had CRS 3763 and picking the GCP on the canvas. In georef configs I choose 3763 as target CRS. The output is added to QGIS with "unknown" CRS (but in the right place)

@nyalldawson after all this seems to be the only outstanding issue. I tried also to georef to WGS84 and the resulting layer shows with "invalid" projection.

**Comment by roya0045** [comment]
> @gioman
> 
> It's always possible that I've made the wrong choice here, and that we should be storing the source **coordinates** instead of pixels in the gcp file! I'm not sure what the original intention was, but we always interpret these values as pixels and the heading states "pixelX" so that's why I went that way.
> 
> But maybe the heading is what's wrong... 😳
> 
> Thinking about a future time when we support vector file georeferencing, it no longer would make sense to use pixels here.
> 
> So in short... I'm 50/50, and would value any insights anyone has!

IMO the coordinates should be the good things to store, as ungeoreferenced images also have their own 'coordinates' and the pixel values are quite specific to a file, but the intent is to move one point to another and from one projection to another. The pixels are just where the point is for a given raster. I could see uses where there is a single gcp files to georeference multiple images, even if the points are out of the image the transformation should still be usable. IE if you have a lot of overlapping images, do a few generic points and then use the same gcps to georeference everything (if they are already georeferenced, from a non-referenced state this wouldn't work).

But as long as it works fine I don't mind it. But I agree that the requirement of this requires juggling with a lot of 'domains' which can be confusing (raster pixels, pts crs, origin, destinations, destination crs, etc.)

**Comment by roya0045** [comment]
@gioman Thanks for testing! I wouldn't have much time at work or out of it to test it in depth as you did, even with a few days of delay.

**Comment by nyalldawson** [comment]
@roya0045 
> IMO the coordinates should be the good things to store

I think I'm leaning that way too (especially given the future vector consideration).

If we go that we we'll definitely need to fix the headers of the .points files though, as they are misleading right now. I'll see what's involved in making this switch tomorrow (and will also re-test with the raster/points from that ticket as a result)

**Comment by roya0045** [comment]
> @roya0045
> 
> > IMO the coordinates should be the good things to store
> 
> I think I'm leaning that way too (especially given the future vector consideration).
> 
> If we go that we we'll definitely need to fix the headers of the .points files though, as they are misleading right now. I'll see what's involved in making this switch tomorrow (and will also re-test with the raster/points from that ticket as a result)

👍 Sure, let me know if I can be of assistance, though it might not be too timely.

File: src/app/georeferencer/qgsgeorefdatapoint.cpp [hunk 1]
```diff
@@ -22,18 +22,17 @@
 #include "qgsgeorefdatapoint.h"
 
 QgsGeorefDataPoint::QgsGeorefDataPoint( QgsMapCanvas *srcCanvas, QgsMapCanvas *dstCanvas,
-                                        const QgsPointXY &pixelCoords, const QgsPointXY &mapCoords,
-                                        const QgsCoordinateReferenceSystem proj, bool enable )
+                                        const QgsPointXY &sourceCoordinates, const QgsPointXY &destinationMapCoords,
+                                        const QgsCoordinateReferenceSystem &destinationCrs, bool enable )
```
Comment (roya0045) [inline src/app/georeferencer/qgsgeorefdatapoint.cpp:26]: I'm hesitant about the name `destinationCrs` , I understand this is the crs of the 'destination' of this point, but it can be confused with the destination crs of the raster.  Could the name simply be `pointCrs` as it belongs to the point and are a property of it, rather than something that should be changed.

**Comment by nyalldawson** [comment]
@gioman @roya0045 

I've re-worked this so that we ALWAYS use source coordinates when saving/loading gcp points files. This fixes the issue described in https://github.com/qgis/QGIS/issues/46901 (that file works without issue now).

I'd love some more testing of this branch before we merge, if you've time!

(Also added a bunch of extra paper cut fixes to this branch, tests, cleaner code, ...)

**Comment by roya0045** [comment]
I'll try to do some test later today, thanks for doing this!

**Comment by gioman** [comment]
@nyalldawson I see weird things.

First case, canvas is in 3857, georef configured to 3763. The destination Y parameter shows with no decimals, double clicking on its cell a lot more show up (but really, on a projected CRS how many are used?), then when moving the focus the value change to one expressed in scientific notation


https://user-images.githubusercontent.com/1951107/153021970-c2e72627-1ef7-4ce2-937f-36429253a8ca.mp4

When canvas is in 4326, georef configured to 3763, the dest x/y values are in a projected crs(?), then if clicking in a cell of the gcp table values change to geographic, but only for dest x.


https://user-images.githubusercontent.com/1951107/153022780-901f1b40-f020-446a-b453-5e8e6f8cebae.mp4

**Comment by roya0045** [comment]
> First case, canvas is in 3857, georef configured to 3763. The destination Y parameter shows with no decimals, double clicking on its cell a lot more show up (but really, on a projected CRS how many are used?), then when moving the focus the value change to one expressed in scientific notation

I think this is out of scope for the fix, the table should still use the default Qt number formatting. What matters is if the numbers are good.

Here is what I was expecting to test when I had the time:

-  [ ] Mixing projected and geographic coords in the input pts
-  [ ] Moving points on the canvas (with the same or a different origin crs)
-  [ ] georefercing an already georeferenced image
-  [ ] Change the destination crs to validate reprojection
-  [ ] (Maybe ) load a gcp file

**Comment by gioman** [comment]
> I think this is out of scope for the fix

maybe, but is totally confusing,

**Comment by roya0045** [comment]
@nyalldawson Did my tests, everything seemed ok, even with a projected dataset.

**Comment by nyalldawson** [comment]
Thanks @roya0045 !

@gioman 
> I see weird things.

These aren't directly impacted by the changes in this PR, but should be fixed by https://github.com/qgis/QGIS/pull/46941
I'd suggest we get this part merged first and then https://github.com/qgis/QGIS/pull/46941 can be reworked on top of master to fix those issues, with appropriate tests added to suit...

**Comment by roya0045** [comment]
I'm curious to see how much rework will be needed in 46941 & #41386  🤣 

Nyall I know you had interest in 41386, let me do the cleanup and you can take over after if you want, no use in having you do some work I could be doing (albeit slower) during the freeze. Other than reworking the dwg importer I didn't find any easy bugs I could fix quickly and willingly at the moment, so I wouldn't mind updating my things in my free-time.

**Comment by nyalldawson** [comment]
Merged for wider testing

**Comment by nyalldawson** [comment]
@gioman 
>  I see weird things.

These should all be fixed by https://github.com/qgis/QGIS/pull/47279

**Comment by gioman** [comment]
> These should all be fixed by #47279

@nyalldawson confirmed!
